src/material/schematics/ng-generate/mdc-migration/rules/template-migration.ts

```diff
@@ -0,0 +1,56 @@
+/**
+ * @license
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import {Migration, ResolvedResource} from '@angular/cdk/schematics';
+import {SchematicContext} from '@angular-devkit/schematics';
+import {StyleMigrator} from './style-migrator';
+import * as compiler from '@angular/compiler';
+
+/**
+ * Traverses the given tree of nodes and runs the given callbacks for each Element node encountered.
+ *
+ * Note that updates to the start tags of html element should be done in the postorder callback,
+ * and updates to the end tags of html elements should be done in the preorder callback to avoid
+ * issues with line collisions.
+ *
+ * @param nodes The nodes of the ast from a parsed template.
+ * @param preorderCallback A function that gets run for each Element node in a preorder traversal.
+ * @param postorderCallback A function that gets run for each Element node in a postorder traversal.
```

Comment:
Usually what we did for migrations that have problems like that (the line collisions), we captured edits and just sorted them in reverse for the offset. That makes things a little easier. Might be worth doing, but not feeling strongly.

Reply:
I considered that option because it can handle more complex queries in a single pass whereas this approach will need two in some cases. However, this solution requires less code which I think is desirable for this project since there are only three components getting migrated

In other words, in this case I'm valuing simplicity over performance